src/violation.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt::Debug;
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+pub trait Violation: Debug + PartialEq + Eq + Serialize + DeserializeOwned {
+    /// The message used to describe the violation.
+    fn message(&self) -> String;
+
+    /// If autofix is (potentially) available for this violation returns another
+    /// function that in turn can be used to obtain a string describing the
+    /// autofix.
+    fn autofix_formatter(&self) -> Option<fn(&Self) -> String> {
+        None
+    }
+
+    /// A placeholder instance of the violation.
+    fn placeholder() -> Self;
+}
+
+/// This trait exists just to make implementing the [`Violation`] trait more
+/// convenient for violations that can always be autofixed.
+pub trait AlwaysAutofixableViolation:
+    Debug + PartialEq + Eq + Serialize + DeserializeOwned
+{
+    /// The message used to describe the violation.
+    fn message(&self) -> String;
+
+    /// The message used to describe the available autofix.
+    fn autofix_description(&self) -> String;
+
+    /// A placeholder instance of the violation.
+    fn placeholder() -> Self;
+}
+
+/// A blanket implementation.
+impl<VA: AlwaysAutofixableViolation> Violation for VA {
+    fn message(&self) -> String {
+        <Self as AlwaysAutofixableViolation>::message(self)
+    }
+
+    fn autofix_formatter(&self) -> Option<fn(&Self) -> String> {
```

Comment:
Is this name required to differ from `autofix_description`, to avoid a collision?

Reply:
No that's not necessary, trait method calls can be qualified like `<MyStruct as MyTrait>.foobar()`. However since the function does something different, I really think it should be called something different.